doc/decisions/library_split.md

```diff
@@ -0,0 +1,40 @@
+# Library Split
+
+## Problem
+
+Only libelektra-core is supposed to access private data but this contradicts the goal to keep the library minimal.
+`kdbprivate.h` was too generic, it contained many other parts next to the struct definitions of Key/KeySet.
+
+## Constraints
+
+## Assumptions
+
+- The [C99 standard, section 5.2.4.1](http://www.open-std.org/jtc1/sc22/wg14/) gives following limit:
+  4095 external identifiers in one translation unit
+
+## Considered Alternatives
+
+- keep current situation
+
+## Decision
+
+Also allow `libelektra-operations` library to access private Key/KeySet.
+Put struct definitions of Key/KeySet in a separate header file, which gets
+included by parts that need it
+
+## Rationale
+
+- allows various users (plugins, applications) to link to (more or less) exactly what they need
+- allows symbol versioning on different levels (for different evolving libraries)
+- allows alternative implementation of parts of Elektra, e.g. a libcore written in Rust
+- facilitates code reuse between plugins
+
+## Implications
+
+- we need to clearly communicate which plugins/libraries must be exactly in the version of the libelektra-core
```

Comment:
I would simply say that any guarantees for API/ABI stability, backwards compatibility, etc. only hold, if all `libelektra-*` libraries are built from the same commit of the repository. If you need another configuration for some reason, you're on your own.

For plugins (`libelektraplugin-*`) things are different. Ideally they only use public API, but in practice they probably need to define which versions are compatible. In theory this could also be enforced via the plugin contract.

Reply:
see #4223